contracts/0.8.25/utils/Confirmations.sol

```diff
@@ -0,0 +1,229 @@
+     *    - Counts existing confirmations that are not expired, i.e. expiry is not exceeded
+     *
+     * 3. Confirmation Management:
+     *    - If all members of the specified roles have confirmed:
+     *      a. Clears all confirmations for this call
+     *      b. Returns true to indicate that the function can be executed
+     *    - If not enough confirmations:
+     *      a. Stores the current confirmations
+     *      b. Returns false to indicate that the function cannot be executed yet
+     *    - Thus, if the caller has all the roles, returns true immediately
+     *
+     * 4. Gas Optimization:
+     *    - Confirmations are stored in a deferred manner using a memory array
+     *    - Confirmation storage writes only occur if the function cannot be executed immediately
+     *    - This prevents unnecessary storage writes when all confirmations are present,
+     *      because the confirmations are cleared anyway after the function is executed,
+     *    - i.e. this optimization is beneficial for the deciding caller and
+     *      saves 1 storage write for each role the deciding caller has
+     *
+     * @param _calldata msg.data of the call (selector + arguments)
+     * @param _roles Array of role identifiers that must confirm the call in order to execute it
+     * @return bool True if all required confirmations are present and the function can be executed, false otherwise
+     *
+     * @notice Confirmations past their expiry are not counted and must be recast
+     * @notice Only members of the specified roles can submit confirmations
+     * @notice The order of confirmations does not matter
+     *
+     */
+    function _collectAndCheckConfirmations(bytes calldata _calldata, bytes32[] memory _roles) internal returns (bool) {
+        if (_roles.length == 0) revert ZeroConfirmingRoles();
+
+        uint256 numberOfRoles = _roles.length;
+        uint256 numberOfConfirms = 0;
+        bool[] memory deferredConfirms = new bool[](numberOfRoles);
+        bool isRoleMember = false;
+
+        ConfirmationStorage storage $ = _getConfirmationsStorage();
+        uint256 expiryTimestamp = block.timestamp + $.confirmExpiry;
+
+        for (uint256 i = 0; i < numberOfRoles; ++i) {
+            bytes32 role = _roles[i];
+            if (_isValidConfirmer(role)) {
+                isRoleMember = true;
+                numberOfConfirms++;
+                deferredConfirms[i] = true;
+
+                emit RoleMemberConfirmed(msg.sender, role, block.timestamp, expiryTimestamp, msg.data);
+            } else if ($.confirmations[_calldata][role] >= block.timestamp) {
+                numberOfConfirms++;
+            }
+        }
+
+        if (!isRoleMember) revert SenderNotMember();
+
+        if (numberOfConfirms == numberOfRoles) {
+            for (uint256 i = 0; i < numberOfRoles; ++i) {
+                bytes32 role = _roles[i];
+                delete $.confirmations[_calldata][role];
```

Comment:
There may be a missing event here to indicate when the required number of confirmations is reached and previous confirmations are cleared